galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,311 @@
+			scope.Debugf("received service for the current, known version: %v", event)
+			return
+		}
+
+		// Store the service.
+		p.services[fullName] = entry
+
+		// Get the associated endpoints, if available.
+		var endpoints *coreV1.Endpoints
+		if endpointsEntry, ok := p.endpoints[fullName]; ok {
+			endpoints = endpointsEntry.Item.(*coreV1.Endpoints)
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(entry, endpoints)
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+
+	case resource.Deleted:
+		// Delete the Service and ServiceEntry
+		delete(p.services, fullName)
+		p.deleteMcpResource(fullName)
+		p.updateVersion()
+	default:
+		scope.Errorf("unknown event kind: %v", event.Kind)
+	}
+}
+
+func (p *Handler) handleEndpointsEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the version has changed.
+		curEntry, ok := p.endpoints[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("received endpoints for the current, known version: %v", event)
+			return
+		}
+
+		p.endpoints[fullName] = entry
+
+		// Look up the service associated with the endpoints.
+		svcEntry, ok := p.services[fullName]
+		if !ok {
+			// Wait until we get a Service before we create the ServiceEntry.
+			scope.Debugf("received endpoints before service for: %s", fullName)
+			return
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(svcEntry, entry.Item.(*coreV1.Endpoints))
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+
+	case resource.Deleted:
+		// The lifecycle of the ServiceEntry is bound to the service, so only delete the endpoints entry here.
+		delete(p.endpoints, fullName)
+
+		// Look up the service associated with the endpoints.
+		svcEntry, ok := p.services[fullName]
+		if !ok {
+			return
+		}
+
+		// Update the MCP entry to clear out the endpoints.
+		mcpEntry, ok := p.toMcpResource(svcEntry, nil)
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+	default:
+		scope.Errorf("unknown event kind: %v", event.Kind)
+	}
+}
+
+func (p *Handler) setMcpEntry(fullName resource.FullName, mcpEntry *mcp.Resource) {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	p.mcpResources[fullName] = mcpEntry
+}
+
+func (p *Handler) deleteMcpResource(fullName resource.FullName) {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	delete(p.mcpResources, fullName)
+}
+
+func (p *Handler) updateVersion() {
+	p.version++
+	monitoring.RecordStateTypeCount(collection.String(), len(p.mcpResources))
+
+	scope.Debugf("in-memory state has changed:\n%v\n", p)
+	p.pendingEvents++
+	p.notifyChanged()
+}
+
+func (p *Handler) notifyChanged() {
+	p.listener.CollectionChanged(collection)
```

Comment:
Where do we make the distinction in if we send the full list of ServiceEntries or just a single ServiceEntry with endpoints changed? This is what @costinm calls "incremental EDS".

Reply:
There is no distinction at this level. The "incremental" part will be handled by "incremental MCP", which is the serving layer for the snapshots generated here.

cc @ayj